### PR TITLE
HPCC-28722 Provide a program for analysing roxie complete: lines

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: tools/roxie/extract-roxie-timings.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    parser.add_argument("--all", "-a", help="Combine all services into a single result", action='store_true')
+    parser.add_argument("--nosummary", "-n", help="Avoid including a summary", action='store_true')
+    parser.add_argument("--summaryonly", "-s", help="Only generate a summary", action='store_true')
+    args = parser.parse_args()
+    combineServices = args.all
+    suppressDetails = args.summaryonly
+    reportSummary = not args.nosummary or args.summayonly
+
+    csv.field_size_limit(0x100000)
+    with open(args.filename) as csv_file:
```
Comment (mckellyln) [inline tools/roxie/extract-roxie-timings.py:52]: I had to add:
    with open(args.filename, encoding='latin1') as csv_file:
to get it to work with another log file, due to some chars causing an error:
  File "/usr/lib64/python3.6/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xa8 in position 7872: invalid start byte
Not sure yet which char or why.

Comment (mckellyln) [reply tools/roxie/extract-roxie-timings.py:52]: I think there are some no-printable chars in the log file - no idea why

File: tools/roxie/extract-roxie-timings.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                            if matchMS:
+                                castValue = (float(matchMS.group(1)) * 60 + float(matchMS.group(2))) * 1000
+                            elif value.endswith("ns"):
+                                castValue = float(value[0:-2])/1000000
+                            elif value.endswith("us"):
+                                castValue = float(value[0:-2])/1000
+                            elif value.endswith("ms"):
+                                castValue = float(value[0:-2])
+                            elif value.endswith("s"):
+                                castValue = float(value[0:-1])*1000
```
Comment (mckellyln) [inline tools/roxie/extract-roxie-timings.py:111]: Another one,
TimeTotalExecute=1h04m15s
ValueError: could not convert string to float: '1h04m15'

File: tools/roxie/extract-roxie-timings.py (hunk 1 above)
Comment (mckellyln) [reply tools/roxie/extract-roxie-timings.py:52]: I get a feeling one of the
SYS:
or
NIC:
system stats log lines might have a rogue char in it ??  I will look more later.

Comment (mckellyln) [reply tools/roxie/extract-roxie-timings.py:52]: Could leave this as default utf-8 encoding but then add a try/catch ?

**Review by mckellyln — COMMENTED** [review]
A couple of comments in-line.

**Comment by ghalliday** [comment]
@mckellyln added a few more interesting changes, and fixed issues you reported.

**Review by mckellyln — COMMENTED** [review]
One last exception:
```
Traceback (most recent call last):
  File "/mnt/disk1/home/kellma02/roxcomplete.py", line 146, in <module>
    castValue = (float(matchHMS.group(1)) * 24 * 60 + float(matchHMS.group(2)) * 60 + float(matchMS.group(3))) * 1000
AttributeError: 'NoneType' object has no attribute 'group'
```
I will get input line ...

**Comment by mckellyln** [comment]
Think this is the line/row ?
```
['0001C164', 'PRG', '2022-12-13', '15:22:09.622', '80877', '86035', '[10.173.117.5:9876{2014}] COMPLETE: didville.ran_bestinfo_batch_service - from 10.176.142.152 complete in 34314 msecs memory=60 Mb priority=-2 agentsreply=17891816 duplicatePackets=4 resentPackets=4 resultsize=11219 continue=0 WhenFirstRow=71229-07-10T20:43:12.172Z TimeTotalExecute=1h05m38s NumRowsProcessed=1017458 NumIndexSeeks=140916 NumIndexScans=120910 NumIndexWildSeeks=103111 NumIndexMerges=48 NumLeafCacheHits=93307 NumNodeCacheHits=57392 NumLeafCacheAdds=10815 NumNodeCacheAdds=1128 NumIndexAccepted=72191 NumIndexRejected=70910 NumIndexRowsRead=129012 TimeSortElapsed=735.973us NumGroups=25961 TimeLocalExecute=56.134s NumAllocations=153168757 SizeAgentReply=17.062Mb TimeAgentWait=7.181s TimeNodeLoad=56.604ms TimeLeafLoad=4.338s TimeLeafLoad=1.209s TimeNodeRead=24.747ms TimeLeafRead=1.473s TimeLeafRead=16.405ms NumNodeDiskFetches=69 NumLeafDiskFetches=5065 TimeNodeFetch=17.610ms TimeLeafFetch=1.437s TimeLeafFetch=16.178ms fDMetaphone1={ NumStarts=280 NumStops=280 TimeLocalExecute=402.145us }']
```

**Comment by ghalliday** [comment]
I didn't change the last call to use matchHMS instead of matchMS


________________________________

**Comment by mckellyln** [comment]
Looks good.

**Review by mckellyln — APPROVED** [review]
Approved

**Pull request merged**